.travis.yml

```diff
@@ -1,12 +1,77 @@
 language: cpp
 dist: trusty
 
-# In order to catch as many weird edge cases in the code with -Werror as
-# possible, we want to test a large range of old and new compilers on both
-# Linux and macOS. This gives us the best possible coverage while maintaining
-# compatibility with a large number of compilers.
-matrix:
+stages:
+    - c-cpp
+    - python
+
+env:
+    global:
+        # Uploading wheels to PyPI, TWINE_PASSWORD is set to a PyPI API token in Travis settings
+        - TWINE_USERNAME=__token__
+
+.testing: &testing
+    stage: c-cpp
+    
+    before_install:
+      - eval "${MATRIX_EVAL}"
+    
+    install:
+      - if [ $TRAVIS_OS_NAME == "linux" ]; then sudo apt-get install valgrind; fi
+    
+    before_script:
+      # Build C/C++ library and apps.
+      - mkdir -p build && cd build && CXXFLAGS="-Werror" cmake -GNinja .. && ninja -v && cd ..
+    
+    script:
+      # Test C/C++ library using CTest.
+      - ninja -C build -v test && cat build/Testing/Temporary/LastTest.log
+      # Check for memory leaks. Returns 2 if there was a memory leak/error, otherwise returns runTests exit code,
+      # which is 0 if all went fine or 1 if some of the tests failed. I test for memory leaks only on linux because
+      # osx returns errors from system libraries, which I would have to supress.
+      # This is run only on smaller number of tests, since executing on valgrind is slow.
+      - if [ $TRAVIS_OS_NAME == "linux" ]; then valgrind --quiet --error-exitcode=2 --tool=memcheck --leak-check=full build/bin/runTests 2; fi
+
+.wheel_build: &wheel_build
+    stage: pywheels
```

Comment:
This should be renamed to python? Let's do the same for the anchor above.

Reply:
Done